### PR TITLE
optimizing `matrix` and `sparse_matrix` for `Prod`

### PR DESCRIPTION
**Pull request by albi3ro**
When we have no overlapping wires between the factors of `Prod`, we can simply use a Kronecker product to combine the matrices.

This provides some pretty sizable speedups.  When testing locally for simple use cases, I'm actually seeing faster matrices for `Prod` than the equivalent `Tensor`.

This PR could work in tandem with other performance improvements like #3038, #3031, #3030 .

### CODE REVIEW

**Review by AmintorDusko — COMMENTED** [review]
Hi @albi3ro! Nice work so far!

**Comment by AmintorDusko** [comment]
Hi @albi3ro, could you please provide some numbers for the speed-ups observed?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -7,8 +7,8 @@ on:
 
 
 env:
```
Comment (AlbertMitjans) [inline .github/workflows/tests.yml:9]: What are all these changes?

Should we do this in another PR?

File: doc/introduction/operations.rst [hunk 2]
```diff
@@ -19,6 +19,7 @@ These operators can be used in quantum functions, like shown in the following ex
         qml.RZ(x, wires=0)
         qml.CNOT(wires=[0,1])
         qml.RY(y, wires=1)
+        qml.T(wires=0).inv()
```
Comment (AlbertMitjans) [inline doc/introduction/operations.rst:22]: Why is this added?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -169,29 +169,6 @@
          [ 0.162,  0.477]])
   ```
 
-* Added the possibility to compute general Renyi entropies in the `ClassicalShadow` class.
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:172]: Why are all these removed?

**Comment by AmintorDusko** [comment]
Could you please double-check your [merge conflict commit](https://github.com/PennyLaneAI/pennylane/pull/3040/commits/fa1484258ccd28ad19a59be72e4df78d35e5318e), and see if it does what you intended it to do?

**Review by AlbertMitjans — DISMISSED** [review]
There are a lot of weird removals and additions in this PR that should be checked.

Moreover, it would be really nice if you could merge this PR with PR #3031 (or wait until it is merged to master) to make sure you have the correct changes.

**Comment by AmintorDusko** [comment]
Hi @AlbertMitjans, the weird changes come from the merge conflict commit. That's why I asked @albi3ro to double-check it.

**Comment by albi3ro** [comment]
@AmintorDusko @AlbertMitjans  Merge conflicts resolved. Sorry about that.

**Comment by AmintorDusko** [comment]
As it is now:
For Sparse Matrices, defining the Hamiltonian as:
```
H = qml.op_sum(
    qml.s_prod(0.5, qml.PauliX(0)),
    qml.s_prod(1.2, qml.prod(qml.PauliY(1), qml.PauliX(0))),
    qml.s_prod(3.4, qml.prod(qml.PauliZ(3), qml.PauliZ(2), qml.PauliY(0))),
    qml.s_prod(2.3, qml.prod(qml.PauliZ(4), qml.PauliY(5)))
)
```
I'm detecting a slowdown in comparison with the most recent PL-master.
image
This branch:
image

For dense operators, the timings seem to be statistically the same:
```
op=qml.prod(qml.PauliX(0), qml.PauliY(1), qml.PauliZ(3), qml.PauliX(4), qml.PauliY(5))
t = qml.operation.Tensor(*op.factors)
```
Most recent master:
image

This branch:
image

**Comment by AmintorDusko** [comment]
Hi. Following our discussion, I re-did the benchmarks in fresh environments and now I can see the nice increase in performance coming for `matrix`.
Pennylane-master:
image
This branch:
image
Great work on that!

Similarly to before I still observe a slowdown for `sparse_matrix`.
With the most recent master and branch:
Pennylane-master:
image
This branch:
image

If you could confirm that would be great.

**Comment by AmintorDusko** [comment]
Hi, again @albi3ro.
I re-did the benchmarks for the `sparse_matrix` implementation with `timeit` and it shows improvement too.
Pyinstrument was probably a not good option for benchmarking this.
Anyway, for PL-master:
image
For this branch:
image

Nice work on that!

**Review by AmintorDusko — APPROVED** [review]

File: doc/releases/changelog-dev.md [hunk 4]
```diff
@@ -393,6 +393,10 @@
   the `wire_order` to the end to avoid computing unnecessary sums and products of huge matrices.
   [(#3030)](https://github.com/PennyLaneAI/pennylane/pull/3030)
 
+* The the factors of `Prod` don't share any wires, the matrix and sparse matrix are computed using
```
Comment (AmintorDusko) [inline doc/releases/changelog-dev.md:396]:
```suggestion
* When the factors of `Prod` don't share any wires, the matrix and sparse matrix are computed using
```

**Review by rmoyard — APPROVED** [review]
Thank you @albi3ro, it looks good 💯 great to see the time improvement by @AmintorDusko !

File: pennylane/ops/op_math/prod.py [hunk 5]
```diff
@@ -205,30 +209,79 @@ def eigvals(self):
...
+
+        if base_label is not None:
+            if isinstance(base_label, str) or len(base_label) != len(self.factors):
+                raise ValueError(
+                    "Prod label requires ``base_label`` keyword to be same length"
+                    " as product factors."
+                )
+            return "@".join(
+                _label(f, decimals, lbl, cache) for f, lbl in zip(self.factors, base_label)
+            )
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:282]:
```suggestion
```

File: pennylane/ops/op_math/prod.py [hunk 6]
```diff
@@ -205,30 +209,44 @@ def eigvals(self):
...
+
+            return math.expand_matrix(reduced_mat, prod_wires, wire_order=wire_order)
+        mats = (op.sparse_matrix() for op in self)
+        full_mat = reduce(sparse_kron, mats)
+        return math.expand_matrix(full_mat, self.wires, wire_order=wire_order)
 
-        wire_order = wire_order or self.wires
 
-        return math.expand_matrix(reduced_mat, prod_wires, wire_order=wire_order)
+        return "@".join(_label(f, decimals, None, cache) for f in self.factors)
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:249]:
```suggestion
```

File: pennylane/ops/op_math/prod.py [hunk 7]
```diff
@@ -17,9 +17,13 @@
 """
 import itertools
 from copy import copy
+from functools import reduce
 from itertools import combinations
 from typing import List, Tuple, Union
 
+import numpy as np
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:24]:
```suggestion
```

**Pull request merged**